soroban-env-host/src/auth.rs

```diff
@@ -0,0 +1,940 @@
+        Ok(RecordedAuthPayload {
+            address: self.address.clone(),
+            invocation: self.root_authorized_invocation.to_xdr_non_metered()?,
+            nonce: self.nonce,
+        })
+    }
+
+    // Checks if there is at least one authorized invocation in the current call
+    // stack.
+    fn has_authorized_invocations_in_stack(&self) -> bool {
+        self.invocation_id_in_call_stack.iter().any(|i| i.is_some())
+    }
+
+    fn invocation_to_xdr(&self, budget: &Budget) -> Result<xdr::AuthorizedInvocation, HostError> {
+        self.root_authorized_invocation.to_xdr(budget)
+    }
+
+    fn push_frame(&mut self) {
+        self.invocation_id_in_call_stack.push(None);
+    }
+
+    fn pop_frame(&mut self) {
+        self.invocation_id_in_call_stack.pop();
+    }
+
+    // Consumes nonce if the nonce check is still needed.
+    // Returns nonce if it has been consumed.
+    // Note, that for the invoker nonce is never needed.
+    fn maybe_consume_nonce(&mut self, host: &Host) -> Result<Option<u64>, HostError> {
+        if !self.need_nonce {
+            return Ok(None);
+        }
+        self.need_nonce = false;
+        if let Some(addr) = &self.address {
+            Ok(Some(host.read_and_consume_nonce(
+                &self.root_authorized_invocation.contract_id,
+                addr,
+            )?))
+        } else {
+            Ok(None)
+        }
+    }
+
+    // Walks a path in the tree defined by `invocation_id_in_call_stack` and
+    // returns the last visited authorized node.
+    fn last_authorized_invocation_mut(&mut self) -> Option<&mut AuthorizedInvocation> {
+        for i in 0..self.invocation_id_in_call_stack.len() {
+            if self.invocation_id_in_call_stack[i].is_some() {
+                return self
+                    .root_authorized_invocation
+                    .last_authorized_invocation_mut(&self.invocation_id_in_call_stack, i + 1);
+            }
+        }
+        None
+    }
+
+    // Tries to match the provided invocation to the authorized sub-invocation
+    // of the current tree and push it to the call stack.
+    // Returns `true` if the match has been found.
+    fn maybe_push_matching_invocation_frame(
+        &mut self,
+        contract_id: &Hash,
+        function_name: Symbol,
+        args: &ScVec,
+    ) -> bool {
+        let mut frame_index = None;
+        if let Some(curr_invocation) = self.last_authorized_invocation_mut() {
+            for i in 0..curr_invocation.sub_invocations.len() {
+                let sub_invocation = &mut curr_invocation.sub_invocations[i];
+                if !sub_invocation.is_exhausted
+                    && &sub_invocation.contract_id == contract_id
+                    && sub_invocation.function_name == function_name
+                    && &sub_invocation.args == args
+                {
+                    frame_index = Some(i);
+                    sub_invocation.is_exhausted = true;
```

Comment:
Are there any weird edge cases that can result from having duplicate sub_invocations? The frame index would be overwritten here and a sub_invocation will become exhausted even if it isn't used. So a contract that calls `authorize` from the same account with the same args twice would never work right? I'm guessing we should break here.

Reply:
Nice catch, there is really no reason not to break here; that's just a bug.